### PR TITLE
Simplify dispatching of sync/async handlers

### PR DESCRIPTION
**Pull request by purple52**
Model asynchronous handler methods as returning Void to reduce the complexity of the handler registry.

### CODE REVIEW

File: core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java [hunk 1]
```diff
@@ -2,52 +2,107 @@
...
-            throw new InvalidHandlerException("Handles method must have exactly one parameter. Found " + handlerMethod.getParameterTypes().length);
+        if (method.getParameterTypes().length != 1) {
+            throw new InvalidHandlerException(
+                    format("Handles method must have exactly one parameter; found %d",
+                            method.getParameterTypes().length));
+        }
+
+        final Class<?> paramClass = method.getParameterTypes()[0];
+        if (paramClass != Envelope.class) {
+            throw new IllegalArgumentException("Handler methods must receive Envelope as an argument");
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java]: is it worth giving he value of paramClass in the exception message?

File: core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java [hunk 2]
```diff
@@ -2,52 +2,107 @@
...
+        }
+
+        if (isVoid(expectedReturnType) && !isVoid(method.getReturnType())) {
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java]: Repeating this action multiple times (including the association at bottom), is it worth just setting the value and reusing?

File: core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java [hunk 3]
```diff
@@ -1,47 +1,37 @@
...
-        return asynchronousMethods.containsKey(name);
-    }
-
-    public boolean canHandleSynchronous(final String name) {
-        return synchronousMethods.containsKey(name);
+    public HandlerMethod get(final String name, final boolean isSynchronous) {
+        if (canHandle(name, isSynchronous)) {
+            return handlerMethods.get(name);
+        } else {
+            throw new MissingHandlerException("No handler registered to handle action: " + name);
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java]: want to add the synchronous state to the error message?

File: core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java [hunk 4]
```diff
@@ -55,29 +45,22 @@ public void register(final Object handlerInstance) {
      * Registers handler instance and method.
      *
      * @param handler handler instance to register.
-     * @param method handler method to register.
+     * @param method  handler method to register.
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java]: extra space

File: core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java [hunk 5]
```diff
@@ -55,29 +45,22 @@ public void register(final Object handlerInstance) {
...
-        }
-    }
-
-    private <T extends HandlerMethod> void put(final Object handler,
-                                               final Method method,
-                                               final BiFunction<Object, Method, T> constructor,
-                                               final Map<String, T> map) {
-
-        T handlerMethod = constructor.apply(handler, method);
+        HandlerMethod handlerMethod = new HandlerMethod(handler, method, method.getReturnType());
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java]: should these 2 local variables be final as there is no intent to update them

File: core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java [hunk 6]
```diff
@@ -35,41 +36,27 @@
...
-        createRegistry(new TestCommandHandlerWithSynchronousHandler());
-        assertThat(registry.canHandleSynchronous(COMMAND_NAME), equalTo(true));
-        assertThat(registry.canHandleAsynchronous(COMMAND_NAME), equalTo(false));
-    }
-
-    @Test
     public void shouldReturnTheRegisteredAsynchronousHandler() {
         createRegistry(new TestCommandHandler());
-        assertThat(registry.getAsynchronous(COMMAND_NAME), notNullValue());
+        assertThat(registry.get(COMMAND_NAME, false), notNullValue());
```
Comment (matt-rich) [inline core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java]: would if be worth moving this into a constant of SYNCHRONOUS = TRUE and doing !SYNCHRONOUS for false for clarity of the test?

File: core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java (hunk 1 above)
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java]: Yep - will add this.

File: core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java (hunk 2 above)
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java]: Error handling still needs to go somewhere! I'd prefer to keep it in one place. Going to try and simplify.

File: core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java (hunk 3 above)
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java]: Done.

File: core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java (hunk 5 above)
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/handler/registry/HandlerRegistry.java]: Done

File: core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java (hunk 6 above)
Comment (purple52) [inline core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java]: Yep - going to add constants to the HandlerMethod class.

**Pull request merged**